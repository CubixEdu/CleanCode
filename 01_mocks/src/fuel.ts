export class Fuel {

    public fill(litersOfFuel: number) {
        return litersOfFuel;
    }
}