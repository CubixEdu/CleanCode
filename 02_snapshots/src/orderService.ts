import { UserService } from "./userService";

export class OrderService {

    constructor(private userService: UserService) { }

    public getUsersOfTheLastHour() {
        const users = [];
        for (let index = 1; index < 5; index++) {
            const user = this.userService.getUserById(index);
            users.push(user);
        }
        return users;
    }
}