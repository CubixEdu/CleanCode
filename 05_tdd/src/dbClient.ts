export class DbClient {

    public getUserNames() {
        return ["batman", "superman", "wonderwoman"];
    }
}