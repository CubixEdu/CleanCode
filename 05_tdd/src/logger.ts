export class Logger {

    public logError(errorMessage: string) {
        console.log(errorMessage);
    }
}