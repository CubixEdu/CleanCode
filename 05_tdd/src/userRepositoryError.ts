export class UserRepositoryError extends Error {

    constructor(message: string, private error: Error) {
        super(message);
    }
}