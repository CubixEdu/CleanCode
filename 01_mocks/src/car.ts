import { Fuel } from "./fuel";

export class Car {

    constructor(private fuel: Fuel) { }

    public drive() {
        this.fuel.fill(35);
    }
}