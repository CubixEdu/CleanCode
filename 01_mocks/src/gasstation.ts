import { Fuel } from "./fuel";

export class GasStation {

    constructor(private fuel: Fuel) { }

    public fillUpCars(numberOfCars: number) {
        for (let index = 1; index < numberOfCars + 1; index++) {
            this.fuel.fill(index + 1);
        }
    }
}