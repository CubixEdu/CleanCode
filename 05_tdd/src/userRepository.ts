import { DbClient } from "./dbClient";
import { Logger } from "./logger";
import { UserRepositoryError } from "./userRepositoryError";

export class UserRepository {

    constructor(private dbClient: DbClient, private logger: Logger) {}

    public getUserNames() {
        try {
            return this.dbClient.getUserNames();
        } catch (error) {
            const errorMessage = 'DbClient failed.';
            this.logger.logError(errorMessage)
            throw new UserRepositoryError(errorMessage, error as Error);
        }
    }
}