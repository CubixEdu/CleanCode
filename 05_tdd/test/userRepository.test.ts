import { mock } from 'jest-mock-extended';
import { DbClient } from '../src/dbClient';
import { UserRepository } from '../src/userRepository';
import { UserRepositoryError } from '../src/userRepositoryError';
import { Logger } from '../src/logger';

const mockedDatabaseClient = mock<DbClient>();
const mockedLogger = mock<Logger>();

describe('UserRepository', () => {
    let sut: UserRepository;

    beforeEach(() => {
        sut = new UserRepository(mockedDatabaseClient, mockedLogger);
    })

    it('should return all the usernames', () => {
        // Arrange     
        const mockUserNames = ["spider man", "captain marvel"];
        mockedDatabaseClient.getUserNames.mockReturnValue(mockUserNames);

        // Act
        const actualResult = sut.getUserNames();

        // Assert
        expect(actualResult).toMatchSnapshot();
    })

    it('should throw a custom error if there is any error thrown from the db client', () => {
        // Arrange
        const error = new Error('DB connection failed.');
        mockedDatabaseClient.getUserNames.mockImplementation(() => { throw error });
        const customErrorMessage = 'DbClient failed.';

        const expectedError = new UserRepositoryError(customErrorMessage, error);

        // Act and assert
        expect(() => { sut.getUserNames() }).toThrow(expectedError);
        expect(mockedLogger.logError).toHaveBeenCalledTimes(1);
        expect(mockedLogger.logError).toHaveBeenCalledWith(customErrorMessage);
    })
})