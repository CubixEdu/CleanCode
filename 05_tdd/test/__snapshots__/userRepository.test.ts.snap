// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRepository should return all the usernames 1`] = `
[
  "spider man",
  "captain marvel",
]
`;
